subprojects/tooling-api/src/main/java/org/gradle/tooling/LongRunningOperation.java

```diff
@@ -120,6 +121,21 @@ public interface LongRunningOperation {
      */
     LongRunningOperation addJvmArguments(String... jvmArguments);
 
+    /**
+     * Sets additional system properties for the build.
+     * <p>
+     * This method has "setter" behavior, so the last invocation will overwrite previously set values.
+     * <p>
+     * Note that there are system properties, like {@code file.encoding}, that Gradle considers immutable for the build. Immutable system properties cannot be modified
+     * with this method. The complete list of immutable properties depends on the target Gradle version. See the {@code org.gradle.process.internal.JvmOptions} class for the details.
+     *
+     * @param systemProperties the system properties add to the Gradle process
+     * @return this
+     * @since 7.6
+     */
+    @Incubating
+    LongRunningOperation withSystemProperties(Map<String, String> systemProperties);
```

Comment:
Thinking about this more, I think we should try to make this work on older Gradle releases too.

I think we can accomplish that by...

- Translating `withSystemProperties(...)` to `-Dfoo=bar` and clearing/restoring System.getProperties() when using a ProviderConnection
- When using a ProviderConnection, clearing/restoring System.getProperties like we are with the embedded executer. 

There might be some other clever way of patching this.

Reply:
Extracted to https://github.com/gradle/gradle/issues/21930